src/genotyper.cc

```diff
@@ -108,15 +108,47 @@ class FormatFieldHelper : public IFormatFieldHelper {
         return ans;
     }
 
+    Status get_default_value(T* val) {
+        if (field_info.default_type == DefaultValueFiller::ZERO) {
+            *val = 0;
+        } else if (field_info.default_type == DefaultValueFiller::MISSING) {
+            switch (field_info.type) {
+                case RetainedFieldType::INT:
+                    *val = bcf_int32_missing;
+                    break;
+                case RetainedFieldType::FLOAT:
+                {
+                    float tmp = 0;
```

Comment:
The C++ compiler generates two different versions of this function at compile time, one with `T` as `int32_t` the other with `T` as `float`. Even though only one branch of the switch can execute at runtime in each copy, both branches have to type-check at compile time. I bet this is the source of the type error (`val` being an `int32_t` in one version and `float` in the other).

[Explicit template specialization](http://en.cppreference.com/w/cpp/language/template_specialization) (case 4, member function of a class template) is the general solution to this type of problem.

However, in this specific case, the `union` trick [used in htslib](https://github.com/samtools/htslib/blob/b6aa0e6e2350ca7836c275082f1d680fdca9e335/htslib/vcf.h#L827) is not bad and probably simpler...